ddtrace/contrib/datadog_lambda/patch.py

```diff
@@ -0,0 +1,89 @@
+from importlib import import_module
+import os
+import signal
+
+from ddtrace.internal.wrapping import unwrap
+from ddtrace.internal.wrapping import wrap
+
+
+class ImpendingTimeout(Exception):
+    pass
+
+
+def crash_flush(_, __):
+    """Raises an `ImpendingTimeout` error which causes the tracer to shutdown."""
+    # TODO: Update to tag error in local root parent instead of raising the error.
+
+    raise ImpendingTimeout("Datadog detected an Impending Timeout")
+
+
+def check_timeout(context):
+    """
+    Creates a timeout to detect when an AWS Lambda handler's remaining
+    time is about to end.
+
+    Crashes flushes when the signal is activated
+    """
+    signal.signal(signal.SIGALRM, crash_flush)
+    remaining_time_in_millis = context.get_remaining_time_in_millis()
+    apm_flush_deadline = int(os.environ.get("DD_APM_FLUSH_DEADLINE", 0))
+    if apm_flush_deadline and apm_flush_deadline <= remaining_time_in_millis:
+        remaining_time_in_millis = apm_flush_deadline
+
+    remaining_time_in_millis = (remaining_time_in_millis - 100) / 1000
```

Comment:
`signal.setitimer(...)` accepts timers in `seconds`, but `remaining_time_in_millis` is given in milliseconds. We substract 100 milliseconds to ensure the traces are flushed before the Lambda function times out and crashes.

Nonetheless, I think I have to change the name of the variable to `remaining_time_in_seconds` and send it to the timer.

Reply:
Ok yeah it was the variable name that was confusing me. Do we know that 100 milliseconds will always be enough? Is there a different way we could ensure that the traces are flushed?

Reply:
It's an estimate really. We will never be sure how long would it take for us to flush spans. A user could have any number of unfinished spans, maybe 100 ms is not enough for the amount of unfinished spans. Maybe in the future we could calculate how long does it take in average to flush a huge amount of unfinished spans. So, to answer the question, it will not always be enough. But in average it might be for the meantime.

I don't think we have another way of making sure that the traces are flushed, since AWS kills the process once the timeout is reached. So not yet.